### ISSUE TITLE
`Iterator[str]` becomes unknown

### ISSUE THREAD

**Issue by parched** [microsoft/pyright#4557]
**Describe the bug**
In the excerpt below the type of `line` becomes unknown for some reason.
When I remove some of the loops or `if` it becomes known.

**To Reproduce**
```python
from typing import Iterator

for _ in ["1"]:
    old_lines: Iterator[str] = iter(["2", "3"])

    try:
        while True:
            line = next(old_lines)
            count = 1
            if count:
                while not (line := next(old_lines)):
                    pass

                while True:
                    if not (line := next(old_lines)):
                        pass
                    elif line.startswith(""):
                        print(line.removeprefix(""))
                    else:
                        old_lines = iter(
                            [line] + list(old_lines))
                        break

    except StopIteration:
        pass
```

**Expected behavior**
I expected no errors, even without any type hints.

**Screenshots or Code**
Weirdly, Pylance correctly infers the type of `line` as `str`, but if I add the type hint anyway, the error goes away.
![image](https://user-images.githubusercontent.com/5975405/215916233-6ede2501-ba4b-46e0-820a-b43c2a5895ba.png)

![image](https://user-images.githubusercontent.com/5975405/215916583-33acf5d1-6018-4f37-8c6e-f49ae3533db2.png)

**VS Code extension or command-line**
Python 3.10
Pylance language server 2023.1.40 (pyright e96b3ece)

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.292, which I just published. It will also be included in a future release of pylance.
